lib/types/src/index.ts

```diff
@@ -1,16 +1,78 @@
 import type { Logger } from 'winston'
 import { Schema, SchemaOutput } from './schema'
+import fs from 'fs'
+import path from 'path'
+import semver from 'semver'
 
-export abstract class Task<O extends Schema = Record<string, never>> {
+const packageJsonPath = path.resolve(__dirname, '../package.json')
+const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'))
+const version: string = packageJson.version
+
+// uses Symbol.for, not Symbol, so they're compatible across different
+// @dotcom-tool-kit/types instances
+
+// used as the name for the property we use to identify classes
+const typeSymbol = Symbol.for('@dotcom-tool-kit/types')
+
+// used to identify the Base, Task and Hook classes
+const baseSymbol = Symbol.for('@dotcom-tool-kit/types/base')
+const taskSymbol = Symbol.for('@dotcom-tool-kit/types/task')
+const hookSymbol = Symbol.for('@dotcom-tool-kit/types/hook')
+
+abstract class Base {
+  static version = version
+  version = version
+
+  static get [typeSymbol](): symbol {
+    return baseSymbol
+  }
+
+  get [typeSymbol](): symbol {
+    return baseSymbol
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  static is<T extends Base>(objectToCheck: any): objectToCheck is T {
+    return objectToCheck[typeSymbol] === this[typeSymbol]
+  }
+
+  static isCompatible<T extends Base>(objectToCheck: unknown): objectToCheck is T {
+    if (!this.is(objectToCheck)) {
+      return false
+    }
+
+    // objectToCheck from a plugin is compatible with this CLI if its version
+    // is semver-compatible with the @dotcom-tool-kit/types included
+    // by the CLI (which is what's calling this). so, prepend ^ to
+    // our version, and check our version satisfies that.
+
+    // this lets e.g. a CLI that includes types@2.2.0 load any plugin
+    // that depends on any higher minor version of types.
+    const range = `^${this.version}`
+    return semver.satisfies(objectToCheck.version, range)
```

Comment:
suggestion: if this returns false, it would also be helpful to return a message like `expected ${objectToCheck.name} version to satisfy ${range}, but got ${objectToCheck.version}`so that users can see what version they need to install.